bee-common/bee-common/src/packable/mod.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! A module that provides a `Packable` trait to serialize and deserialize types.
+
+extern crate alloc;
+
+mod error;
+#[cfg(feature = "io")]
+mod io;
+mod packer;
+mod unpacker;
+
+use alloc::{boxed::Box, vec::Vec};
+use core::convert::Infallible;
+
+pub use error::{UnknownTagError, UnpackError};
+pub use packer::{Packer, VecPacker};
+pub use unpacker::{SliceUnpacker, UnexpectedEOF, Unpacker};
+
+pub use bee_common_derive::Packable;
+
+/// A type that can be packed and unpacked.
+///
+/// Almost all basic sized types implement this trait. This trait can be derived using the
+/// `bee_common_derive::Packable` macro. If you need to implement this trait manually, use the provided
+/// implementations as a guide.
+pub trait Packable: Sized {
+    /// The error type that can be returned if some semantic error occurs while unpacking.
+    ///
+    /// It is recommended to use `core::convert::Infallible` if this kind of error cannot happen or
+    /// `UnpackEnumError` when implementing this trait for an enum.
+    type Error;
+    /// Pack this value into the given `Packer`.
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error>;
+    /// Unpack this value from the given `Unpacker`.
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>>;
+    /// The size of the value in bytes after being packed.
+    fn packed_len(&self) -> usize;
+}
+
+macro_rules! impl_packable_for_int {
+    ($ty:ty) => {
+        impl Packable for $ty {
+            type Error = Infallible;
+
+            fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+                packer.pack_bytes(&self.to_le_bytes())
+            }
+
+            fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
+                let mut bytes = [0u8; core::mem::size_of::<Self>()];
+                unpacker.unpack_bytes(&mut bytes).map_err(UnpackError::Unpacker)?;
```

Comment:
Do we really need the map ?

Reply:
we do not!

Reply:
Maybe do a pass and see if we can remove some other ?